drivers/media/i2c/mira016.inl

```diff
@@ -0,0 +1,6501 @@
+			// Debug print
+			printk(KERN_INFO "[MIRA016]: mira016_write_target_frame_time_reg target_frame_time_us = %u.\n",
+				   mira016->target_frame_time_us);
+			printk(KERN_INFO "[MIRA016]: width %d, hblank %d, vblank %d, height %d, ctrl->val %d.\n",
+				   mira016->mode->width, mira016->mode->hblank, mira016->mode->min_vblank, mira016->mode->height, ctrl->val);
+			ret = mira016_write_target_frame_time_reg(mira016, mira016->target_frame_time_us);
+			break;
+		case V4L2_CID_HBLANK:
+			printk(KERN_INFO "[MIRA016]: V4L2_CID_HBLANK CALLED = %d.\n",
+				ctrl->val);
+			break;
+		default:
+			dev_info(&client->dev,
+					 "ctrl(id:0x%x,val:0x%x) is not handled\n",
+					 ctrl->id, ctrl->val);
+			ret = -EINVAL;
+			break;
+		}
+	}
+
+	pm_runtime_put(&client->dev);
+
+	// TODO: FIXIT
+	return ret;
+}
+
+static int mira016_s_ctrl(struct v4l2_ctrl *ctrl)
+{
+	struct mira016 *mira016 =
+		container_of(ctrl->handler, struct mira016, ctrl_handler);
+	struct i2c_client *client = v4l2_get_subdevdata(&mira016->sd);
+	int ret = 0;
+
+	// printk(KERN_INFO "[MIRA016]: mira016_s_ctrl() id: %X value: %X.\n", ctrl->id, ctrl->val);
+
+	/* Previously, register writes when powered off will be buffered.
+	 * The buffer will be written to sensor when start_streaming.
+	 * Now, register writes happens immediately, even powered off.
+	 * Register writes when powered off will fail.
+	 * Users need to make sure first power on then write register.
+	 */
+
+	switch (ctrl->id)
+	{
+	case AMS_CAMERA_CID_MIRA_REG_W:
+		ret = mira016_v4l2_reg_w(mira016, ctrl->val);
+		break;
+	default:
+		dev_info(&client->dev,
+				 "set ctrl(id:0x%x,val:0x%x) is not handled\n",
+				 ctrl->id, ctrl->val);
+		ret = -EINVAL;
+		break;
+	}
+
+	// TODO: FIXIT
+	return ret;
+}
+
+static int mira016_g_ctrl(struct v4l2_ctrl *ctrl)
+{
+	struct mira016 *mira016 =
+		container_of(ctrl->handler, struct mira016, ctrl_handler);
+	struct i2c_client *client = v4l2_get_subdevdata(&mira016->sd);
+	int ret = 0;
+
+	// printk(KERN_INFO "[MIRA016]: mira016_g_ctrl() id: %X.\n", ctrl->id);
+
+	/*
+	 * Ideally, V4L2 register read should happen only when powered on.
+	 * However, perhaps there are use cases that,
+	 * reading other I2C addr is desired when mira sensor is powered off.
+	 * Therefore, the check of "powered" flag is disabled for now.
+	 */
+
+	switch (ctrl->id)
+	{
+	case AMS_CAMERA_CID_MIRA_REG_R:
+		ret = mira016_v4l2_reg_r(mira016, (u32 *)&ctrl->cur.val);
+		ctrl->val = ctrl->cur.val;
+		break;
+	default:
+		dev_info(&client->dev,
+				 "get ctrl(id:0x%x) is not handled\n",
+				 ctrl->id);
+		ret = -EINVAL;
+		break;
+	}
+
+	// TODO: FIXIT
+	return ret;
+}
+
+static const struct v4l2_ctrl_ops mira016_ctrl_ops = {
+	.s_ctrl = mira016_set_ctrl,
+};
+
+static const struct v4l2_ctrl_ops mira016_custom_ctrl_ops = {
+	.g_volatile_ctrl = mira016_g_ctrl,
+	.s_ctrl = mira016_s_ctrl,
+};
+
+/* list of custom v4l2 ctls */
+static struct v4l2_ctrl_config custom_ctrl_config_list[] = {
```

Comment:
If this is being used to allow userspace to poke random registers, then please look at implementing `s_register` and `g_register` in `v4l2_subdev_core_ops`. It also gets around your need to set and cache the register address through Write in order to be able to implement read.